python/ray/serve/tests/unit/test_router.py

```diff
@@ -489,5 +495,199 @@ async def test_max_queued_requests_updated(
+            DeploymentID("a", "b"),
+            "random",
+            Mock(),
+            FakeCounter(tag_keys=("deployment", "route", "application")),
+            FakeGauge(tag_keys=("deployment", "application")),
+        )
+
+        # Not an autoscaling deployment, should not push metrics
+        assert not metrics_manager.should_send_scaled_to_zero_optimized_push(0)
+
+        # No queued requests at the handle, should not push metrics
+        metrics_manager.deployment_config = DeploymentConfig(
+            autoscaling_config=AutoscalingConfig()
+        )
+        assert not metrics_manager.should_send_scaled_to_zero_optimized_push(0)
+
+        # Current number of replicas is non-zero, should not push metrics
+        metrics_manager.inc_num_queued_requests()
+        assert not metrics_manager.should_send_scaled_to_zero_optimized_push(1)
+
+        # All 3 conditions satisfied, should push metrics
+        assert metrics_manager.should_send_scaled_to_zero_optimized_push(0)
+
+    @patch(
+        "ray.serve._private.router.RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE", "1"
+    )
+    def test_push_autoscaling_metrics_to_controller(self):
+        timer = MockTimer()
+        start = random.randint(50, 100)
+        timer.reset(start)
+        deployment_id = DeploymentID("a", "b")
+        handle_id = "random"
+        mock_controller_handle = Mock()
+
+        with patch("time.time", new=timer.time):
+            metrics_manager = RouterMetricsManager(
+                deployment_id,
+                handle_id,
+                mock_controller_handle,
+                FakeCounter(tag_keys=("deployment", "route", "application")),
+                FakeGauge(tag_keys=("deployment", "application")),
+            )
+            metrics_manager.deployment_config = DeploymentConfig(
+                autoscaling_config=AutoscalingConfig()
+            )
+
+            # Set up some requests
+            n = random.randint(0, 5)
+            replica_tags = [get_random_string() for _ in range(3)]
+            running_requests = defaultdict(int)
+            for _ in range(n):
+                metrics_manager.inc_num_queued_requests()
+            for _ in range(20):
+                r = random.choice(replica_tags)
+                running_requests[r] += 1
+                metrics_manager.inc_num_running_requests_for_replica(r)
+
+            # Check metrics are pushed correctly
+            metrics_manager.push_autoscaling_metrics_to_controller()
+            mock_controller_handle.record_handle_metrics.remote.assert_called_with(
+                deployment_id=deployment_id,
+                handle_id=handle_id,
+                queued_requests=n,
+                running_requests=running_requests,
+                send_timestamp=start,
+            )
+
+    @patch(
+        "ray.serve._private.router.RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE", "1"
+    )
+    @patch("ray.serve._private.router.MetricsPusher", new=Mock)
```

Comment:
please follow up to remove this mock and instead either (1) make a fake implementation or (2) directly test with the real implementation